keps/sig-node/4112-passdown-resources-to-cri/README.md

```diff
@@ -0,0 +1,1124 @@
+**Note:** Generally we also wait at least two releases between beta and
+GA/stable, because there's no opportunity for user feedback, or even bug reports,
+in back-to-back releases.
+
+**For non-optional features moving to GA, the graduation criteria must include
+[conformance tests].**
+
+[conformance tests]: https://git.k8s.io/community/contributors/devel/sig-architecture/conformance-tests.md
+
+#### Deprecation
+
+- Announce deprecation and support policy of the existing flag
+- Two versions passed since introducing the functionality that deprecates the flag (to address version skew)
+- Address feedback on usage/changed behavior, provided on GitHub issues
+- Deprecate the flag
+-->
+
+#### Alpha
+
+- Feature implemented behind a feature flag
+- Initial unit tests completed and enabled
+
+#### Beta
+
+- Gather feedback from developers and surveys
+- Feature gate enabled by default
+- containerd and CRI-O runtimes have released versions that have adopted the
+  new CRI API changes
+- The [NRI API](https://github.com/containerd/nri) has adopted the feature
+
+#### GA
+
+- No bugs reported in the previous cycle
+
+### Upgrade / Downgrade Strategy
+
+<!--
+If applicable, how will the component be upgraded and downgraded? Make sure
+this is in the test plan.
+
+Consider the following in developing an upgrade/downgrade strategy for this
+enhancement:
+- What changes (in invocations, configurations, API use, etc.) is an existing
+  cluster required to make on upgrade, in order to maintain previous behavior?
+- What changes (in invocations, configurations, API use, etc.) is an existing
+  cluster required to make on upgrade, in order to make use of the enhancement?
+-->
+
+The feature gate (in kubelet) controls the feature enablement. Existing runtime
+implementations will continue to work as previously, even if the feature is
+enabled.
+
+### Version Skew Strategy
+
+<!--
+If applicable, how will the component handle version skew with other
+components? What are the guarantees? Make sure this is in the test plan.
+
+Consider the following in developing a version skew strategy for this
+enhancement:
+- Does this enhancement involve coordinating behavior in the control plane and
+  in the kubelet? How does an n-2 kubelet without this feature available behave
+  when this feature is used?
+- Will any other components on the node change? For example, changes to CSI,
+  CRI or CNI may require updating that component before the kubelet.
+-->
+
+The feature is node-local (kubelet-only) so there is no dependencies or effects
+to other Kubernetes components.
+
+The behavior is unchanged if either kubelet or the CRI runtime running on a
+node does not support the feature. If kubelet has the feature enabled but the
+CRI runtime does not support it, the CRI runtime will ignore the new fields in
+the CRI API and function as previously. Similarly, if the CRI runtime supports
+the feature but the kubelet does not, the runtime will resort to the previous
+behavior.
```

Comment:
Even if the kubelet and container runtime support the feature.. and in some cases even if the NRI plugins are installed...  long running pods created N-1 or N-2 etc. will be a problem for the NRI plugins that expect to see the additional updating pod event and/or additional pod metadata.  A restart of the plugin will be synching some pods that have the new metadata and some that do not.

Reply:
Gut tells me at some point we're going to need a version tag for the contents of the pod metadata. Sure we can continue down the non-breaking add changes only path.  But I'm reminded of certain alpha's that did not make it and/or redrafted completely (DRA).  Thus begging the question.

Reply:
If the kubelet and the runtime support this, then a long running pod (that gets restarted) should get this information from the kubelet. I'm a bit thin on the exact details here but updating the kubelet and the runtime without draining the node may have this issue.

@mikebrow, thoughts, do you have suggestions to address this?

Reply:
Yes draining the node.. Stopping and restarting the pods is probably the only reasonable way to guarantee the new metadata is attached to a running pod when the NRI plugin syncs with the container runtime.

Was just pointing out the need to mention that for this feature. Turn the feature off.. no metadata, turn it back on.. only new pods will have it. yada yada

Reply:
K. I now added a note to the **Does enabling the feature change any default behavior** section under PRR questionnaire:

```
Long running pods that persist (without restart) over kubelet and CRI runtime
update which enables the feature may experience version skew of the metadata.
After enabling the feature, the CRI runtime does not have the aggregated
information of all resources of the pod, provided with this feature, as the
kubelet didn't restart these pods (didn't send the CreatePodSandbox CRI
request). This may affect some scenarios e.g. NRI plugins. This "metadata skew"
can be avoided by draining the node before updating the kubelet and the CRI
runtime.
```

WDYT @mikebrow ?

Reply:
works..